src/Security/MemberAuthenticator/LoginHandler.php

```diff
@@ -0,0 +1,247 @@
+    {
+        Session::clear('SessionForms.MemberLoginForm.Email');
+        Session::clear('SessionForms.MemberLoginForm.Remember');
+
+        $member = Security::getCurrentUser();
+        if ($member->isPasswordExpired()) {
+            return $this->redirectToChangePassword();
+        }
+
+        // Absolute redirection URLs may cause spoofing
+        $backURL = $this->getBackURL();
+        if ($backURL) {
+            return $this->redirect($backURL);
+        }
+
+        // If a default login dest has been set, redirect to that.
+        $defaultLoginDest = Security::config()->get('default_login_dest');
+        if ($defaultLoginDest) {
+            return $this->redirect($defaultLoginDest);
+        }
+
+        // Redirect the user to the page where they came from
+        if ($member) {
+            // Welcome message
+            $message = _t(
+                'SilverStripe\\Security\\Member.WELCOMEBACK',
+                'Welcome Back, {firstname}',
+                ['firstname' => $member->FirstName]
+            );
+            Security::singleton()->setLoginMessage($message, ValidationResult::TYPE_GOOD);
+        }
+
+        // Redirect back
+        return $this->redirectBack();
+    }
+
+    /**
+     * Try to authenticate the user
+     *
+     * @param array $data Submitted data
+     * @param ValidationResult $result
+     * @return Member Returns the member object on successful authentication
+     *                or NULL on failure.
+     */
+    public function checkLogin($data, &$result)
+    {
+        $member = $this->authenticator->authenticate($data, $result);
+        if ($member instanceof Member) {
+            return $member;
+        }
+        // No member, can't login
+        $this->extend('authenticationFailed', $data);
+
+        return null;
+    }
+
+    /**
+     * Try to authenticate the user
+     *
+     * @param Member $member
+     * @param array $data Submitted data
+     * @param HTTPRequest $request
+     * @return Member Returns the member object on successful authentication
+     *                or NULL on failure.
+     */
+    public function performLogin($member, $data, $request)
+    {
+        /** IdentityStore */
+        Injector::inst()->get(IdentityStore::class)->logIn($member, !empty($data['Remember']), $request);
+
+        return $member;
+    }
+
+    /**
+     * Invoked if password is expired and must be changed
+     *
+     * @skipUpgrade
+     * @return HTTPResponse
+     */
+    protected function redirectToChangePassword()
+    {
+        $cp = ChangePasswordForm::create($this, 'ChangePasswordForm');
+        $cp->sessionMessage(
+            _t('SilverStripe\\Security\\Member.PASSWORDEXPIRED', 'Your password has expired. Please choose a new one.'),
+            'good'
+        );
+        $changedPasswordLink = Security::singleton()->Link('changepassword');
+
+        return $this->redirect($this->addBackURLParam($changedPasswordLink));
+    }
+}
```

Comment:
I'm happy for this method to be moved up the inheritance hierarchy to RequestHandler.

(Copypasta stuff)